# Week 01 "Pre-lecture" HW

## 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [13]:
import pandas as pd

# URL of the dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

# Import the dataset
data = pd.read_csv(url)

# Display the first few rows of the dataset
print("First few rows of the dataset:")
print(data.head())

# Check for missing values
missing_values = data.isnull().sum()

# Display missing values
print("\nMissing values in each column:")
print(missing_values)


First few rows of the dataset:
   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  

Missing values in each column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
clas

## 2. Use the code provided in the ChatBot session to print out the number of rows and columns of the dataset (df.shape)

In [2]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Get the number of rows and columns
num_rows, num_columns = df.shape

print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")


Number of rows: 891
Number of columns: 15


**My own general definitions of the meaning of "observations" and "variables."**
- Observations: Rows in a dataset. Each row is about one person/item/incident or a specific thing being studied.
- Variables: Columns in a dataset. Columns are about the characteristics/features of a row.

## 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [46]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Summary statistics for numeric columns
print(df.describe())

# set the max columns to none
pd.set_option('display.max_columns', None)


         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


In [43]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Count of unique values in the 'class' column
print(df['class'].value_counts())


class
Third     491
First     216
Second    184
Name: count, dtype: int64


## 4. Explain the discrepancies between the size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column
(a) The Titanic dataset has both numeric and non-numeric columns, so df.describe() will display fewer columns than the total number reported by df.shape. This is because df.describe() only summarizes numeric columns. Only 6 columns are displayed by df.describe(), while a total of 15 columns are reported by df.shape.

(b) The Titanic dataset has missing values, so the count for each column in df.describe() will be less than the total number of rows reported by df.shape. This is because count only includes non-null values, while df.shape shows the total number of rows, including those with null values. For example, the count for "age" is 714 while the number of columns given by df.shape is 891, and that is because of the 177 missing values in this column.


## 5. My own paraphrasing summarization of the difference between an "attribute" and a "method"
- An "attribute" is like a characteristic piece of information about something, while a "method" is like an action that the object can do.
- "Attributes" store data, and methods define actions or behaviours.

## Summaries of my ChatBot session (01)
Summary of Interaction
1. Understanding Attributes vs. Methods in pandas:
- Attributes: These are properties of an object that are accessed directly without parentheses. They represent the state or characteristics of the object.
 - Example: df.shape provides the dimensions of a pandas DataFrame, and df.columns provides the column names.
- Methods: These are functions associated with an object that perform operations or computations. Methods are called with parentheses, and they may take arguments.
 - Example: df.describe() computes summary statistics for numeric columns in a DataFrame, and df.head(n) returns the first n rows of the DataFrame.

2. Discrepancies Observed:
- Number of Columns in .shape vs. .describe():
 - .shape returns the total number of rows and columns in the DataFrame.
 - .describe() summarizes only numeric columns, so it might show fewer columns than .shape if there are non-numeric columns.
- Count of Non-Null Values in .describe() vs. Rows in .shape:
 - The count in .describe() shows the number of non-null entries for each numeric column. If there are missing values, this count will be less than the total number of rows reported by .shape.
- Code Examples:
 - Using .describe():

In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Summary statistics for numeric columns
print(df.describe())


- Using df['column'].value_counts():

In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Count of unique values in the 'class' column
print(df['class'].value_counts())


4. In-Depth Explanation:
- Attributes: Characteristics or properties of an object accessed directly. Examples include df.shape and df.columns.
- Methods: Functions that perform actions or computations, called with parentheses. Examples include df.describe() and df.head().

**Link to the chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/286193ee-5873-4f55-abc3-38df5047c214

# "Post-lecture" HW 

## 6. The definitions of each of the summary statistics provided by the df.describe() method.
- Count: The number of non-null (non-missing) values in the column.

- Mean: The average value of the data points in the column. It is calculated as the sum of all values divided by the count of non-null values.

- Standard Deviation (std): A measure of the spread or dispersion of the data. It quantifies how much the values in the column deviate from the mean. A higher standard deviation indicates that the data points are spread out over a wider range.

- Minimum (min): The smallest value in the column.

- 25% (First Quartile): This is the value below which 25% of the data points lie. It is also known as the first quartile (Q1).

- 50% (Median): This is the value that separates the dataset into two equal halves. It is the midpoint of the data, also known as the median.

- 75% (Third Quartile): This is the value below which 75% of the data points lie. It is also known as the third quartile (Q3).

- Maximum (max): The largest value in the column.

When df.describe() is used on a DataFrame, it automatically excludes missing values from the calculations. This behaviour applies to each individual column in the DataFrame. 

## 7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words
1. **Provide an example of a "use case" in which using df.dropna() might be preferred over using del df['col']:** Let's say there is a dataset that has some missing values across the rows. These missing values are not concentrated in a specific column, instead, they are all spread out to different columns. Notice that the number of rows that have missing values is not significant compared to the total size of the dataset. With the given situation, using df.dropna(axis=0) to get rid of the rows with missing values might be preferred because losing a small portion of the dataset is acceptable, and all of the columns are safe.

2. **Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna():** Let's suppose there is a dataset that has many columns and one of them is irrelevant and has many missing values. Other than this column, the rest of the columns are clean. In this situation, it's preferred to use del df['col'] to take out the only column that is problematic, without affecting the rest.

3. **Discuss why applying del df['col'] before df.dropna() when both are used together could be important:** First, when both del df['col'] and df.dropna() are used together it's important to apply del df['col'] first because it's more efficient. By manually choosing which column to get rid of, the work for df.dropna() to do becomes easier. Second, applying del df['col'] first can avoid unnecessary data loss. Recall the example in question 2, if applying df.dropna() before del df['col'], many rows will be dropped because of the particular column that has many missing values.

4. **Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.**

In [2]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)


In [4]:
# Remove the 'deck' column due to excessive missing data
del df['deck']


In [5]:
# Remove rows with any remaining missing data
df_cleaned = df.dropna()


In [6]:
# Show the cleaned dataset
df_cleaned.info()

# Display the first few rows
print(df_cleaned.head())


<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     712 non-null    int64  
 1   pclass       712 non-null    int64  
 2   sex          712 non-null    object 
 3   age          712 non-null    float64
 4   sibsp        712 non-null    int64  
 5   parch        712 non-null    int64  
 6   fare         712 non-null    float64
 7   embarked     712 non-null    object 
 8   class        712 non-null    object 
 9   who          712 non-null    object 
 10  adult_male   712 non-null    bool   
 11  embark_town  712 non-null    object 
 12  alive        712 non-null    object 
 13  alone        712 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 73.7+ KB
   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1   

## 7.4 Justification 
**Step 1:** I removed the 'deck' column using del df['col']. The reason why I chose to delete this column is that it has a high percentage of missing values (688/891 ≈ 0.77). Using del df['col'] to remove this column first can avoid unnecessary data loss, which could have happened if using df.dropna() first.

**Step 2:** After cleaning the dataset with del df['col'], I removed any row that still has missing values using df.dropna(). The dataset should not have any missing values after this step.

## 7.4 Before and After Report
***Before:***

In [4]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


- The dataset had 891 rows and 15 columns.
- The dataset had 177 missing values in the 'age' column; 2 missing values in the 'embarked' column; 688 missing values in the 'deck' column.

***After:***

In [8]:
df_cleaned.info()


<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     712 non-null    int64  
 1   pclass       712 non-null    int64  
 2   sex          712 non-null    object 
 3   age          712 non-null    float64
 4   sibsp        712 non-null    int64  
 5   parch        712 non-null    int64  
 6   fare         712 non-null    float64
 7   embarked     712 non-null    object 
 8   class        712 non-null    object 
 9   who          712 non-null    object 
 10  adult_male   712 non-null    bool   
 11  embark_town  712 non-null    object 
 12  alive        712 non-null    object 
 13  alone        712 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 73.7+ KB


- The dataset now has 712 rows and 14 columns.
- There is no missing value in the dataset because every column has exactly 712 non-null entries.

## Summary of my ChatBot session (02)
1. Difference Between df.dropna() and del df['col']:
- df.dropna() removes rows or columns containing missing values, depending on the axis argument.
- del df['col'] deletes a specific column by name, regardless of whether it contains missing values.

2. Discussion on Efficient Use of df.dropna() and del df['col']:
- Using df.dropna() is preferred when you want to drop rows or columns with missing values while preserving the rest of the structure.
- del df['col'] is useful when removing specific, irrelevant, or problematic columns without considering their missing values.

3. Example Use Case Favoring df.dropna():
- In a house price prediction dataset, df.dropna() is preferred to retain all features (columns) and only drop rows with missing values to avoid losing important information in other columns.

4. Opposite Use Case Favoring del df['col']:
- In a customer database, the "Survey Response" column with 90% missing values can be removed using del df['col'] because the rest of the dataset is clean, and dropping the column ensures that valuable data in other columns is retained.

5. Order of Operations: Why del df['col'] Before df.dropna() Matters:
- Removing irrelevant or problematic columns using del df['col'] first ensures that missing data in those columns doesn’t unnecessarily cause rows to be dropped by df.dropna(). This helps preserve valuable data and ensures a more intentional cleaning process.

6. Applying del df['col'] and df.dropna() on the Titanic Dataset:
- Loaded the Titanic dataset and removed the 'deck' column due to excessive missing values using del df['deck']. Afterward, applied df.dropna() to remove rows with remaining missing data. The cleaned dataset was displayed.

7. Understanding Missing and Non-Missing (Null/Non-Null) Data:
- Null values represent missing data (NaN), and non-null values are valid entries. Pandas uses .isnull() to identify null values and .notnull() to check for non-null values.
- The .info() method helps identify non-null counts in columns and detect missing values.

8. Summary Statistics from df.describe():
- Explanation of summary statistics such as count, mean, std, min, 25%, 50% (median), 75%, and max, and how each provides insight into the dataset's distribution and central tendencies.

This covers all our exchanges, explaining the differences between missing data handling methods, strategies for efficient data cleaning, and deeper insights into null/non-null values and summary statistics.

**Link to this chat log history:** https://chatgpt.com/share/bbbf4dec-53fe-4595-b8ea-3a985836a592

## 8. Give brief explanations in your own words for any requested answers to the questions below
1. **Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you**

For me, df.groupby("col1")["col2"].describe() can be divided into 3 parts to be understood. First, df.groupby("col1") decides what column are we grouping. For example, if I choose to group the "sex" column then there will be two groups: male and female. And then, use ["col2"] to choose another column in the dataset, here, l would like to choose the "survived" column. Finally, .describe(): can calculate statistics for the "survived" column in the "male" group and the "female" group. 

**Here is an example of what I am talking about:**

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

summary_stats = df.groupby("sex")["survived"].describe()

print(summary_stats)


        count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0


In this example, I first grouped the "sex" column using df.groupby("sex"), and then I chose another column "survived". Finally, .describe() helped me calculate the statistics for the "survived" column within the sex of each passenger.

2. **Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?**

This is because the count in df.groupby("col1")["col2"].describe() gives information about the number of non-null values (rows/observations) in "col2" for each group of "col1". This means that it actually shows the distribution of data across groups of "col1". The count in df.describe(), on the other hand, simply shows how many non-null values for every column in the data set.

3. **Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT**



***A. Forget to include import pandas as pd in your code:***

In [1]:
# URL of the dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

# Import the dataset
data = pd.read_csv(url)

NameError: name 'pd' is not defined

***B. Mistype "titanic.csv" as "titanics.csv":***

In [9]:
import pandas as pd

# Import the dataset
data = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv')

HTTPError: HTTP Error 404: Not Found

***C. Try to use a dataframe before it's been assigned into the variable:***

In [15]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

Df.groupby("class")["age"].describe()

NameError: name 'Df' is not defined

***D. Forget one of the parentheses somewhere the code:***

In [19]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

df.groupby("class")["age"].describe()

SyntaxError: '(' was never closed (375293931.py, line 5)

***F. Use a column name that's not in your data for the groupby and column selection:***

In [28]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

summary_stats = df.groupby("Sex")["age"].describe()

print(summary_stats)

KeyError: 'Sex'

In [25]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

summary_stats = df.groupby("sex")["Age"].describe()

print(summary_stats)


KeyError: 'Column not found: Age'

***G. Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question:***

In [26]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

summary_stats = df.groupby(sex)["age"].describe()

print(summary_stats)

NameError: name 'sex' is not defined

In [27]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

summary_stats = df.groupby("sex")[age].describe()

print(summary_stats)

NameError: name 'age' is not defined

## 8.3 My opinion of working with ChatBot to troubleshoot and fix the coding errors or with a Google search:
After intentionally introducing the above errors into my code, I think interacting with ChatGPT to troubleshoot is more efficient and easier than using Google to search for and fix errors. All of the errors above (except for error G) were well resolved by ChatGPT. ChatGPT provided me with not only solutions to fixing the errors but also provided me things like examples of correct/incorrect code, common situations leading to the particular Error, and steps to debug. A Google search, on the other hand, was not as comprehensive as ChatGPT. Most solutions I found were on a website called "stackoverflow." There were questions and answers from humans. However, it took time to read the question and try to find a solution from many answers. ChatGPT is more straightforward than a Google search in terms of this aspect. In addition, When trying to troubleshoot the last error G, I found that both approaches did not really work. I believe it is because an error like this ("NameError: name 'age' is not defined") can be caused by many types of mistakes and it is too hard to troubleshoot.

## Summary of my ChatBot session (03)
Here’s a summary of our exchanges, organized by topic, so you can submit it for your homework assignment:

1. Initial Summary Analyses on Titanic Dataset:
- You mentioned that you were working with the Titanic dataset and performing some initial summary analyses. You provided a link to the dataset hosted on GitHub and asked for help understanding the code and the analysis being performed.
- I explained the operation of df.groupby("col1")["col2"].describe(), detailing how it groups the data by col1 and provides descriptive statistics for col2 within each group.

2. Difference Between df.describe() and groupby(...).describe():
- You asked about the difference in how df.describe() and groupby(...).describe() handle missing values and count.
- I explained that df.describe() gives global summary statistics for each column, showing missingness across the entire DataFrame, whereas groupby(...).describe() provides group-wise statistics, showing how data is distributed within each group.

3. Explaining the Primary Function of groupby(...).describe():
- You asked about "another primary function" of count in groupby(...).describe() besides handling missing values.
- I explained that the primary function is to show how data is distributed across groups, which is more important for understanding the balance or imbalance of data in each group. This information is crucial for analyzing group representations and making statistically sound comparisons.

4. Debugging Errors:
- You encountered several errors during your analysis, and I assisted in debugging those errors:
- NameError: name 'pd' is not defined: This happened because pandas was not imported. I explained how to fix it by adding import pandas as pd at the top of the code.
- HTTPError: 404: You were trying to load the dataset from a URL that was not available. I provided the correct URL to load the Titanic dataset and explained how to load it properly.
- NameError: name 'Df' is not defined: This occurred due to a case sensitivity issue with the variable name. I clarified that Python is case-sensitive and suggested using df consistently.
- SyntaxError: '(' was never closed: This error happened because an opening parenthesis was not closed. I explained how to find and fix unclosed parentheses.
- KeyError: 'Column not found: Age': I explained that this error occurs when trying to access a column that does not exist, possibly due to a typo or incorrect capitalization. I suggested using df.columns to check for the exact column names and ensure correct case sensitivity.
- KeyError: 'Sex' and NameError: 'sex' is not defined: These errors arose due to incorrect column references. I explained how to properly reference DataFrame columns using the correct column names and syntax, like df['sex'].
- NameError: 'age' is not defined: I clarified that columns in pandas must always be referenced using the DataFrame object, like df['age'], and not as standalone variables.

5. General Advice on Debugging:
- For each error, I provided specific explanations on how to fix them, emphasizing the importance of case sensitivity, correct column names, and proper referencing of DataFrame objects in pandas.

This summary should provide a comprehensive record of our exchanges, including the technical explanations and debugging assistance related to your homework on the Titanic dataset analysis.

**Link to this chat log history:** https://chatgpt.com/share/71c2bdc5-7592-4641-9e2e-a9e9333feee7

## 9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?
Yes.